File: 程序员面试金典/3.3/StackOfPlates.java
class StackOfPlates {
    private List<Stack<Integer>> stacks;
    private int cap;

    public StackOfPlates(int cap) {
        this.stacks = new ArrayList<>();
        this.cap = cap;
    }
    
    public void push(int val) {
        if (cap == 0)
            return;
        if (stacks.isEmpty() || stacks.get(stacks.size() - 1).size() == cap) {
            Stack<Integer> stack = new Stack<>();
            stack.push(val);
            stacks.add(stack);
        }
        else
            stacks.get(stacks.size() - 1).push(val);
    }
    
    public int pop() {
        return popAt(stacks.size() - 1);
    }
    
    public int popAt(int index) {
        if (index < 0 || index >= stacks.size())
            return -1;
        int val = stacks.get(index).pop();
        if (stacks.get(index).empty())
            stacks.remove(index);
        return val;
    }

}

/**
 * Your StackOfPlates object will be instantiated and called as such:
 * StackOfPlates obj = new StackOfPlates(cap);
 * obj.push(val);
 * int param_2 = obj.pop();
 * int param_3 = obj.popAt(index);
 */

File: 程序员面试金典/3.3/StackOfPlates.py
class StackOfPlates:

    def __init__(self, cap: int):
        self.stacks = []
        self.cap = cap

    def push(self, val: int) -> None:
        if self.cap == 0:
            return
        if not self.stacks or len(self.stacks[-1]) == self.cap:
            self.stacks.append([val])
        else:
            self.stacks[-1].append(val)

    def pop(self) -> int:
        return self.popAt(-1)

    def popAt(self, index: int) -> int:
        try:
            val = self.stacks[index].pop()
            if not self.stacks[index]:
                self.stacks.pop(index)
            return val
        except IndexError:
            return -1


# Your StackOfPlates object will be instantiated and called as such:
# obj = StackOfPlates(cap)
# obj.push(val)
# param_2 = obj.pop()
# param_3 = obj.popAt(index)
